File: project/libaes/src/cipher.cpp
// C
#include <stddef.h>
#include <stdint.h>

// C++
#include <algorithm>
#include <array>
#include <stdexcept>

// libaes
#include <aes/cipher.hpp>
#include <aes/constants.hpp>
#include <aes/primitives.hpp>
#include <aes/utilities.hpp>

/// The size of a round key.
constexpr size_t ROUND_KEY_SIZE= aes::BLOCK_SIZE;

/// The size (in bytes) of the state.
constexpr size_t STATE_WIDTH = 4;
constexpr size_t STATE_HEIGHT = 4;

/// Convert a size in terms of bytes to a size in terms of bits.
constexpr size_t BITS(size_t bytes)
{
    return bytes * aes::BITS_PER_BYTE;
}

/// Convert a size in terms of bytes to a size in terms of words.
constexpr size_t WORDS(size_t bytes)
{
    return bytes / aes::WORD_SIZE;
}

/**
 * Expand a key to generate the various round keys.
 * 
 * @param[in] key        The key to expand.
 * @param[in] key_size   The size (in bytes) of the key.
 * @param[in] num_rounds The number of round keys to create.
 * @param[in] round_keys A buffer to store the round keys.
 * 
 * @c round_keys is expected to be a 2D array of width @c ROUND_KEY_SIZE and 
 * height @c num_rounds.
 */
static void expand_key(const uint8_t key[], 
                       size_t key_size,
                       size_t num_rounds,
                       uint8_t round_keys[]);

/**
 * Add a round key to the current state.
 * 
 * @param[in]      round_key The round key to add to the state.
 * @param[in, out] state     The state to add the round key to.
 */
static void add_round_key(const uint8_t round_key[],
                          uint8_t state[]);

/**
 * Shift each of the rows in the state.
 * 
 * @param[in] state The current state.
 */
static void shift_rows(uint8_t state[]);

/**
 * Mix each of the columns in the state.
 * 
 * @param[in] state The current state.
 */
static void mix_columns(uint8_t state[]);

namespace aes
{
    void cipher_encode(Type type,
                       const uint8_t key[],
                       const uint8_t block[],
                       uint8_t out[])
    {
        if (key == nullptr)   throw std::invalid_argument("key cannot be null");
        if (block == nullptr) throw std::invalid_argument("block cannot be null");
        if (out == nullptr)   throw std::invalid_argument("out cannot be null");

        std::array<uint8_t, STATE_HEIGHT * STATE_WIDTH> state = {};
        std::array<uint8_t, ROUND_KEY_SIZE * MAX_ROUNDS> round_keys = {};

        size_t key_size = KEY_SIZE(type);
        size_t total_rounds = TOTAL_ROUNDS(type);

        // The input block is our initial state.
        aes::transpose(block, STATE_WIDTH, STATE_HEIGHT, state.data());

        // 1. Key Expansion
        expand_key(key, key_size, total_rounds, round_keys.data());

        // 2. Initial Round
        add_round_key(&round_keys[0], state.data());

        // 3. Intermediate Rounds (9, 11, or 13)
        for (int round = 1; round < (total_rounds - 1); round++)
        {
            substitute(state.data(), state.size(), state.data());
            shift_rows(state.data());
            mix_columns(state.data());
            add_round_key(&round_keys[round * ROUND_KEY_SIZE], state.data());
        }

        // 4. Final Round
        substitute(state.data(), state.size(), state.data());
        shift_rows(state.data());
        add_round_key(&round_keys[(total_rounds - 1) * ROUND_KEY_SIZE], state.data());

        // Our resulting state is the ciphertext.
        aes::transpose(state.data(), STATE_WIDTH, STATE_HEIGHT, out);
    }

} // namespace aes

void expand_key(const uint8_t key[], 
                size_t key_size,
                size_t total_rounds,
                uint8_t round_keys[]) 
{
    static const std::array<uint8_t, 10 * aes::WORD_SIZE> ROUND_CONSTANT_BYTES =
    {
        0x01, 0x00, 0x00, 0x00,
        0x02, 0x00, 0x00, 0x00,
        0x04, 0x00, 0x00, 0x00,
        0x08, 0x00, 0x00, 0x00,
        0x10, 0x00, 0x00, 0x00,
        0x20, 0x00, 0x00, 0x00,
        0x40, 0x00, 0x00, 0x00,
        0x80, 0x00, 0x00, 0x00,
        0x1B, 0x00, 0x00, 0x00,
        0x36, 0x00, 0x00, 0x00
    };

    static const auto ROUND_CONSTANTS = reinterpret_cast<const uint32_t*>(ROUND_CONSTANT_BYTES.data());

    auto key_words = reinterpret_cast<const uint32_t*>(key); 
    auto words = reinterpret_cast<uint32_t*>(round_keys);

    size_t num_key_words = WORDS(key_size);
    size_t num_round_key_words = WORDS(ROUND_KEY_SIZE);
    size_t num_words = total_rounds * num_round_key_words;

    for (size_t word_index = 0; word_index < num_words; word_index++)
    {
        if (word_index < num_key_words)
        {
            words[word_index] = key_words[word_index];
        }
        else if ((word_index % num_key_words) == 0)
        {
            words[word_index] = 
                words[word_index - num_key_words]
                ^ aes::substitute(aes::rotate_left(words[word_index - 1]))
                ^ ROUND_CONSTANTS[(word_index / num_key_words) - 1];
        }
        else if ((num_key_words > 6) && ((word_index % num_key_words) == 4))
        {
            words[word_index] = 
                words[word_index - num_key_words]
                ^ aes::substitute(words[word_index - 1]);
        }
        else
        {
            words[word_index] = 
                words[word_index - num_key_words]
                ^ words[word_index - 1];
        }
    }
}

void add_round_key(const uint8_t round_key[],
                   uint8_t state[]) 
{
    for (size_t row = 0; row < STATE_HEIGHT; row++)
    {
        for (size_t column = 0; column < STATE_WIDTH; column++)
        {
            state[(row * STATE_WIDTH) + column] ^= round_key[(column * STATE_WIDTH) + row];
        }
    }
}

void shift_rows(uint8_t state[]) 
{
    auto state_words = reinterpret_cast<uint32_t*>(state);
    for (size_t row = 0; row < STATE_HEIGHT; row++)
    {
        state_words[row] = aes::rotate_left(state_words[row], row);
    }
}

void mix_columns(uint8_t state[]) 
{
    constexpr std::array<std::array<uint8_t, 4>, 4> MDS_MATRIX = 
    {{
        { 2, 3, 1, 1 },
        { 1, 2, 3, 1 },
        { 1, 1, 2, 3 },
        { 3, 1, 1, 2 }
    }};

    for (size_t column = 0; column < STATE_WIDTH; column++)
    {
        // Calculate each element in the column multiplied by the coefficients
        // 1, 2, and 3 in Rijndael's Galois field. Access the results using:
        //
        //     column_g[i][c]
        //
        // Where i is the index of the byte in the column, and c is the 
        // coefficient - 1.
        std::array<std::array<uint8_t, 3>, STATE_HEIGHT> products = {};
        for (size_t i = 0; i < products.size(); i++)
        {
            // column[i] * 1
            products[i][0] = state[(i * STATE_WIDTH) + column];

            // column[i] * 2
            products[i][1] = products[i][0] << 1;
            if (products[i][0] & 0x80)
            {
                products[i][1] ^= 0x1B;
            }

            // column[i] * 3 
            products[i][2] = products[i][0] ^ products[i][1];
        }

        for (size_t row = 0; row < STATE_HEIGHT; row++)
        {
            uint8_t* value = &state[(row * STATE_WIDTH) + column];

            *value = 0;
            for (int i = 0; i < MDS_MATRIX[row].size(); i++)
            {
                // Add element i of the column multiplied by the coefficient 
                // listed in the corresponding element in the matrix row.
                uint8_t coefficient = MDS_MATRIX[row][i];
                *value ^= products[i][coefficient - 1];
            }
        }
    }
}

File: project/pyaes/cipher.py
from util import *

###############################################################################
# Constants
###############################################################################

# The size (in bytes) of an AES block.
BLOCK_SIZE = 16

###############################################################################
# Main Algorithm
###############################################################################


def aes_cipher(data: bytes, key: bytes) -> bytes:
    """Encode a block of data using the AES cipher.

    The variant of AES used (AES-128, AES-196, or AES-256) depends on the size
    of the key.

    Args:
        data: The block of data to encode. Must have a length of 16 bytes.
        key: The key to mix with the data.

    Returns:
        The encoded data.
    """
    if len(data) != BLOCK_SIZE:
        raise ValueError(f'Data length of {len(data)} is invalid'
                         f' - must be {BLOCK_SIZE}')

    state = transpose_flat(data)

    # 1. Key Expansion
    round_keys = list(_expand_key(key))

    # 2. Initial Round
    state = _add_round_key(state, round_keys[0])

    # 3. Intermediate Rounds (9, 11, or 13)
    for round_key in round_keys[1:-1]:
        state = _substitute_word(state)
        state = _shift_rows(state)
        state = _mix_columns(state)
        state = _add_round_key(state, round_key)

    # 4. Final Round
    state = _substitute_word(state)
    state = _shift_rows(state)
    state = _add_round_key(state, round_keys[-1])

    return transpose_flat(state)


###############################################################################
# Primitive Operations
###############################################################################


def _substitute_byte(byte: int) -> int:
    """Substitute a byte using the Rijndael S-box substitution box."""
    if not hasattr(_substitute_byte, 's_box'):
        p = 1
        q = 1

        _substitute_byte.s_box = {0: 0x63}
        while len(_substitute_byte.s_box) < 256:
            # Multiply p by 3
            p ^= (p << 1) ^ (0x11B if bool(p & 0x80) else 0)
            p &= 0xFF

            # Divide q by 3
            q ^= (q << 1) & 0xFF
            q ^= (q << 2) & 0xFF
            q ^= (q << 4) & 0xFF
            q ^= 0x09 if bool(q & 0x80) else 0
            q &= 0xFF

            # Compute the affine transformation.
            _substitute_byte.s_box[p] = \
                q ^ \
                rotate_byte_left(q, 1) ^ \
                rotate_byte_left(q, 2) ^ \
                rotate_byte_left(q, 3) ^ \
                rotate_byte_left(q, 4) ^ \
                0x63

    return _substitute_byte.s_box[byte]


def _substitute_word(word: bytes) -> bytes:
    """Substitute each of the bytes in a word."""
    return bytes([_substitute_byte(byte) for byte in word])


def _reverse_substitute_byte(byte: int) -> int:
    """Reverse a substitution via the Rijndael S-box."""
    if not hasattr(_reverse_substitute_byte, 'reverse_s_box'):
        _reverse_substitute_byte.reverse_s_box = {_substitute_byte(i): i
                                                  for i in range(0xFF)}

    return _reverse_substitute_byte.reverse_s_box[byte]


def _g_mul(a: int, b: int) -> int:
    """Multiply two bytes in Rijndael's Galois field."""
    result = 0

    for i in range(8):
        if (b & 1) == 1:
            result ^= a

        high_bit = a & 0x80
        a = (a << 1) & 0xFF
        if high_bit:
            a ^= 0x1B
        b >>= 1

    return result


def _mix_column(column: bytes) -> bytes:
    """Mix a column of the state."""
    mds_matrix = [[2, 3, 1, 1],
                  [1, 2, 3, 1],
                  [1, 1, 2, 3],
                  [3, 1, 1, 2]]

    return bytes([reduce(xor, (_g_mul(c, x) for c, x in zip(row, column)))
                  for row in mds_matrix])


###############################################################################
# Steps
###############################################################################


def _expand_key(key: bytes) -> List[bytes]:
    """Expand the cipher key to get each of the round keys."""
    round_constants = [
        bytes([0x01, 0x00, 0x00, 0x00]),
        bytes([0x02, 0x00, 0x00, 0x00]),
        bytes([0x04, 0x00, 0x00, 0x00]),
        bytes([0x08, 0x00, 0x00, 0x00]),
        bytes([0x10, 0x00, 0x00, 0x00]),
        bytes([0x20, 0x00, 0x00, 0x00]),
        bytes([0x40, 0x00, 0x00, 0x00]),
        bytes([0x80, 0x00, 0x00, 0x00]),
        bytes([0x1B, 0x00, 0x00, 0x00]),
        bytes([0x36, 0x00, 0x00, 0x00]),
    ]
    
    key_words = [key[i:i+4] for i in range(0, len(key), 4)]
    key_size = len(key_words)

    total_rounds = \
        11 if key_size == 4 else \
        13 if key_size == 6 else \
        15 if key_size == 8 else \
        None

    if not total_rounds:
        raise ValueError(f'Invalid key size: {len(key)}')

    # The first round's key is just the original key.
    words = key_words

    for i in range(len(words), 4 * total_rounds):
        prev_key_word = words[-key_size]
        prev_word = words[-1]

        if (i % key_size) == 0:
            word = xor_words(
                prev_key_word,
                _substitute_word(rotate_word_left(prev_word)),
                round_constants[(i // key_size) - 1])
        elif key_size > 6 and (i % key_size) == 4:
            word = xor_words(
                prev_key_word,
                _substitute_word(prev_word))
        else:
            word = xor_words(
                prev_key_word,
                prev_word)

        words.append(word)

    return [bytes(chain(*words[i:i+4])) for i in range(0, len(words), 4)]


def _add_round_key(state: bytes, key: bytes) -> bytes:
    """Add the round key to the state."""
    return bytes([a ^ b for a, b in zip(state, transpose_flat(key))])


def _shift_rows(state: bytes) -> bytes:
    """Shift the bytes of each row in the state."""
    rows = (state[i:i+4] for i in range(0, len(state), 4))
    shifted_rows = (rotate_word_left(row, amount=i)
                    for i, row in enumerate(rows))
    return bytes(chain(*shifted_rows))


def _mix_columns(state: bytes) -> bytes:
    """Mix each of the columns in the state."""
    rows = [state[i:i+4] for i in range(0, len(state), 4)]
    columns = transpose(rows)
    mixed_columns = [_mix_column(column) for column in columns]
    return bytes(chain(*transpose(mixed_columns)))


###############################################################################
# Debugging
###############################################################################


def print_words(word1, *words):
    words = (word1, *words)
    for i, word in enumerate(words):
        text = f'{word.hex()}'
        if len(words) > 1:
            text = f'[{i}]: {text}'
        print(text)


###############################################################################
# Main
###############################################################################


def main():
    data = b'\0' * 16
    key = b'\0' * 16
    print(aes_cipher(data, key).hex())


if __name__ == '__main__':
    main()
